src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TagHelperAttributes : ReadOnlyTagHelperAttributes<TagHelperAttribute>, IList<TagHelperAttribute>
+    {
+        public TagHelperAttributes()
+            : base()
+        {
+        }
+
+        public TagHelperAttributes(IEnumerable<TagHelperAttribute> attributes)
+            : base(attributes)
+        {
+        }
+
+
+        public new TagHelperAttribute this[int index]
+        {
+            get
+            {
+                return base[index];
+            }
+            [param: NotNull]
+            set
+            {
+                ValidateAttribute(value, nameof(value));
+
+                Attributes[index] = value;
+            }
+        }
+
+        public new TagHelperAttribute this[[NotNull] string key]
+        {
+            get
+            {
+                return base[key];
+            }
+            [param: NotNull]
+            set
+            {
+                // Name will be null if user attempts to set the attribute via an implicit conversion:
+                // output.Attributes["someName"] = "someValue"
```

Comment:
Seems much more straightforward to just create a set-only `this[]` that takes object than just to deal with conversions, Especially if the conversion isn't correct!